File: src/core/flow-parser.ts
/**
 * 流程解析和验证器
 */

import * as yaml from 'js-yaml';
import { readFileSync, writeFileSync, existsSync, readdirSync, statSync } from 'fs';
import { join, basename } from 'path';
import {
  TestFlow,
  FlowNode,
  FlowEdge,
  NodeType
} from './types.js';

// 节点类型验证
const NODE_TYPES: NodeType[] = ['Start', 'End', 'Action', 'Assert', 'Extract', 'Condition', 'SubFlow', 'Wait', 'Navigate'];
const ACTION_METHODS = ['click', 'input', 'select', 'hover', 'scroll', 'wait', 'screenshot'];
const ASSERT_TYPES = ['text', 'visible', 'hidden', 'enabled', 'contains'];

/**
 * 加载流程图文件
 */
export function loadFlow(filePath: string): TestFlow {
  const content = readFileSync(filePath, 'utf-8');
  const ext = filePath.toLowerCase();

  let flow: TestFlow;
  if (ext.endsWith('.yaml') || ext.endsWith('.yml')) {
    flow = yaml.load(content) as TestFlow;
  } else if (ext.endsWith('.json')) {
    flow = JSON.parse(content);
  } else {
    throw new Error(`不支持的文件格式: ${filePath}`);
  }

  // 验证流程
  validateFlow(flow);

  return flow;
}

/**
 * 加载目录下所有流程图
 */
export function loadFlows(dirPath: string): TestFlow[] {
  const flows: TestFlow[] = [];

  if (!existsSync(dirPath)) {
    return flows;
  }

  const files = readdirSync(dirPath);
  for (const file of files) {
    const filePath = join(dirPath, file);
    const stat = statSync(filePath);

    if (stat.isFile() && (file.endsWith('.yaml') || file.endsWith('.yml') || file.endsWith('.json'))) {
      try {
        const flow = loadFlow(filePath);
        flows.push(flow);
      } catch (e) {
        console.warn(`加载流程失败: ${file}`, e);
      }
    }
  }

  return flows;
}

/**
 * 保存流程图到文件
 */
export function saveFlow(flow: TestFlow, filePath: string): void {
  const ext = filePath.toLowerCase();

  let content: string;
  if (ext.endsWith('.yaml') || ext.endsWith('.yml')) {
    content = yaml.dump(flow, { indent: 2, lineWidth: 120 });
  } else if (ext.endsWith('.json')) {
    content = JSON.stringify(flow, null, 2);
  } else {
    throw new Error(`不支持的文件格式: ${filePath}`);
  }

  writeFileSync(filePath, content, 'utf-8');
}

/**
 * 验证流程图
 */
export function validateFlow(flow: TestFlow): { valid: boolean; errors: string[] } {
  const errors: string[] = [];

  // 检查必填字段
  if (!flow.id) errors.push('流程图缺少 id 字段');
  if (!flow.name) errors.push('流程图缺少 name 字段');
  if (!flow.nodes || flow.nodes.length === 0) errors.push('流程图没有节点');
  // 边不是必填的，允许空流程图

  // 检查节点
  const nodeIds = new Set<string>();
  for (const node of flow.nodes || []) {
    if (!node.id) {
      errors.push('节点缺少 id');
      continue;
    }
    if (nodeIds.has(node.id)) {
      errors.push(`节点 id 重复: ${node.id}`);
    }
    nodeIds.add(node.id);

    if (!node.type) {
      errors.push(`节点 ${node.id} 缺少 type`);
    } else if (!NODE_TYPES.includes(node.type)) {
      errors.push(`节点 ${node.id} 类型无效: ${node.type}`);
    } else {
      // 检查各类型节点的必填字段
      switch (node.type) {
        case 'Action':
          // action 配置暂时不做必填校验
          if (!node.action) {
            console.log(`Action 节点 ${node.id} 缺少 action 配置 (可忽略)`);
          } else {
            if (!node.action.method) console.log(`Action 节点 ${node.id} 缺少 method (可忽略)`);
            if (node.action.method && !ACTION_METHODS.includes(node.action.method)) {
              errors.push(`Action 节点 ${node.id} method 无效: ${node.action.method}`);
            }
          }
          break;
        case 'Assert':
          // assert 配置暂时不做必填校验
          if (!node.assert) {
            console.log(`Assert 节点 ${node.id} 缺少 assert 配置 (可忽略)`);
          } else {
            if (!node.assert.type) console.log(`Assert 节点 ${node.id} 缺少 type (可忽略)`);
            if (node.assert.type && !ASSERT_TYPES.includes(node.assert.type)) {
              errors.push(`Assert 节点 ${node.id} type 无效: ${node.assert.type}`);
            }
          }
          break;
        case 'Extract':
          // extract 配置暂时不做必填校验
          if (!node.extract) {
            console.log(`Extract 节点 ${node.id} 缺少 extract 配置 (可忽略)`);
          }
          break;
        case 'Condition':
          // condition 配置暂时不做必填校验
          if (!node.condition) {
            console.log(`Condition 节点 ${node.id} 缺少 condition 配置 (可忽略)`);
          }
          break;
        case 'SubFlow':
          // subFlow 配置暂时不做必填校验
          if (!node.subFlow) {
            console.log(`SubFlow 节点 ${node.id} 缺少 subFlow 配置 (可忽略)`);
          }
      }
    }
  }

  // 检查边的引用
  for (const edge of flow.edges || []) {
    if (!nodeIds.has(edge.source)) {
      errors.push(`边的源节点不存在: ${edge.source}`);
    }
    if (!nodeIds.has(edge.target)) {
      errors.push(`边的目标节点不存在: ${edge.target}`);
    }
  }

  // 检查必须有一个 Start 和 End 节点
  const startNodes = flow.nodes.filter(n => n.type === 'Start');
  const endNodes = flow.nodes.filter(n => n.type === 'End');

  if (startNodes.length === 0) errors.push('缺少 Start 节点');
  if (startNodes.length > 1) errors.push('Start 节点只能有一个');
  if (endNodes.length === 0) errors.push('缺少 End 节点');

  return {
    valid: errors.length === 0,
    errors
  };
}

/**
 * 获取节点的所有出边
 */
export function getOutgoingEdges(flow: TestFlow, nodeId: string): FlowEdge[] {
  return flow.edges.filter(e => e.source === nodeId);
}

/**
 * 获取节点的所有入边
 */
export function getIncomingEdges(flow: TestFlow, nodeId: string): FlowEdge[] {
  return flow.edges.filter(e => e.target === nodeId);
}

/**
 * 根据ID获取节点
 */
export function getNodeById(flow: TestFlow, nodeId: string): FlowNode | undefined {
  return flow.nodes.find(n => n.id === nodeId);
}

/**
 * 查找 Start 节点
 */
export function findStartNode(flow: TestFlow): FlowNode | undefined {
  return flow.nodes.find(n => n.type === 'Start');
}

/**
 * 查找所有 End 节点
 */
export function findEndNodes(flow: TestFlow): FlowNode[] {
  return flow.nodes.filter(n => n.type === 'End');
}

File: backend/src/main/java/com/uitestflow/util/FlowParser.java
package com.uitestflow.util;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import com.fasterxml.jackson.dataformat.yaml.YAMLGenerator;
import com.uitestflow.dto.ValidationResult;
import com.uitestflow.model.*;
import org.springframework.stereotype.Component;

import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.stream.Collectors;

/**
 * 流程解析和验证器
 */
@Component
public class FlowParser {

    private static final Set<NodeType> NODE_TYPES = new HashSet<>(Arrays.asList(
            NodeType.Start, NodeType.End, NodeType.Action, NodeType.Assert,
            NodeType.Extract, NodeType.Condition, NodeType.SubFlow,
            NodeType.AIAction, NodeType.AIQuery, NodeType.AIAssert,
            NodeType.Wait, NodeType.Navigate
    ));

    private static final Set<String> ACTION_METHODS = new HashSet<>(Arrays.asList(
            "click", "input", "select", "hover", "scroll", "wait", "screenshot"
    ));

    private static final Set<String> ASSERT_TYPES = new HashSet<>(Arrays.asList(
            "text", "visible", "hidden", "enabled", "contains"
    ));

    private final ObjectMapper objectMapper = new ObjectMapper();
    private final ObjectMapper yamlObjectMapper;

    public FlowParser() {
        // 配置 JSON 解析器忽略未知字段
        objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);

        // 配置 YAML 工厂
        YAMLFactory yamlFactory = YAMLFactory.builder()
                .disable(YAMLGenerator.Feature.WRITE_DOC_START_MARKER)
                .build();
        yamlObjectMapper = new ObjectMapper(yamlFactory);
        yamlObjectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
    }

    /**
     * 加载流程图文件
     */
    public TestFlow loadFlow(String filePath) throws IOException {
        String content = new String(Files.readAllBytes(Paths.get(filePath)));

        TestFlow flow;
        if (filePath.toLowerCase().endsWith(".yaml") || filePath.toLowerCase().endsWith(".yml")) {
            flow = yamlObjectMapper.readValue(content, TestFlow.class);
        } else if (filePath.toLowerCase().endsWith(".json")) {
            flow = objectMapper.readValue(content, TestFlow.class);
        } else {
            throw new IOException("不支持的文件格式: " + filePath);
        }

        // 验证流程
        validateFlow(flow);

        return flow;
    }

    /**
     * 加载目录下所有流程图
     */
    public List<TestFlow> loadFlows(String dirPath) throws IOException {
        List<TestFlow> flows = new ArrayList<>();

        File dir = new File(dirPath);
        if (!dir.exists() || !dir.isDirectory()) {
            return flows;
        }

        File[] files = dir.listFiles();
        if (files == null) {
            return flows;
        }

        for (File file : files) {
            if (file.isFile() && (file.getName().endsWith(".yaml") ||
                    file.getName().endsWith(".yml") || file.getName().endsWith(".json"))) {
                try {
                    TestFlow flow = loadFlow(file.getAbsolutePath());
                    flows.add(flow);
                } catch (Exception e) {
                    System.out.println("加载流程失败: " + file.getName() + ", " + e.getMessage());
                }
            }
        }

        return flows;
    }

    /**
     * 保存流程图到文件
     */
    public void saveFlow(TestFlow flow, String filePath) throws IOException {
        String content;
        if (filePath.toLowerCase().endsWith(".yaml") || filePath.toLowerCase().endsWith(".yml")) {
            content = yamlObjectMapper.writeValueAsString(flow);
        } else if (filePath.toLowerCase().endsWith(".json")) {
            content = objectMapper.writerWithDefaultPrettyPrinter().writeValueAsString(flow);
        } else {
            throw new IOException("不支持的文件格式: " + filePath);
        }

        Files.write(Paths.get(filePath), content.getBytes());
    }

    /**
     * 验证流程图
     */
    public ValidationResult validateFlow(TestFlow flow) {
        List<String> errors = new ArrayList<>();

        // 检查必填字段
        if (flow.getId() == null || flow.getId().isEmpty()) {
            errors.add("流程图缺少 id 字段");
        }
        if (flow.getName() == null || flow.getName().isEmpty()) {
            errors.add("流程图缺少 name 字段");
        }
        if (flow.getNodes() == null || flow.getNodes().isEmpty()) {
            errors.add("流程图没有节点");
        }

        // 检查节点
        Set<String> nodeIds = new HashSet<>();
        if (flow.getNodes() != null) {
            for (FlowNode node : flow.getNodes()) {
                if (node.getId() == null || node.getId().isEmpty()) {
                    errors.add("节点缺少 id");
                    continue;
                }
                if (nodeIds.contains(node.getId())) {
                    errors.add("节点 id 重复: " + node.getId());
                }
                nodeIds.add(node.getId());

                if (node.getType() == null) {
                    errors.add("节点 " + node.getId() + " 缺少 type");
                } else if (!NODE_TYPES.contains(node.getType())) {
                    errors.add("节点 " + node.getId() + " 类型无效: " + node.getType());
                } else {
                    // 检查各类型节点的配置
                    switch (node.getType()) {
                        case Action:
                            if (node.getAction() != null && node.getAction().getMethod() != null) {
                                if (!ACTION_METHODS.contains(node.getAction().getMethod().name().toLowerCase())) {
                                    errors.add("Action 节点 " + node.getId() + " method 无效: " + node.getAction().getMethod());
                                }
                            }
                            break;
                        case Assert:
                            if (node.getAssert_() != null && node.getAssert_().getType() != null) {
                                if (!ASSERT_TYPES.contains(node.getAssert_().getType().name().toLowerCase())) {
                                    errors.add("Assert 节点 " + node.getId() + " type 无效: " + node.getAssert_().getType());
                                }
                            }
                            break;
                        default:
                            break;
                    }
                }
            }
        }

        // 检查边的引用
        if (flow.getEdges() != null) {
            for (FlowEdge edge : flow.getEdges()) {
                if (!nodeIds.contains(edge.getSource())) {
                    errors.add("边的源节点不存在: " + edge.getSource());
                }
                if (!nodeIds.contains(edge.getTarget())) {
                    errors.add("边的目标节点不存在: " + edge.getTarget());
                }
            }
        }

        // 检查必须有一个 Start 和 End 节点
        if (flow.getNodes() != null) {
            long startCount = flow.getNodes().stream()
                    .filter(n -> n.getType() == NodeType.Start)
                    .count();
            long endCount = flow.getNodes().stream()
                    .filter(n -> n.getType() == NodeType.End)
                    .count();

            if (startCount == 0) errors.add("缺少 Start 节点");
            if (startCount > 1) errors.add("Start 节点只能有一个");
            if (endCount == 0) errors.add("缺少 End 节点");
        }

        if (errors.isEmpty()) {
            return ValidationResult.success();
        } else {
            return ValidationResult.failure(errors);
        }
    }

    /**
     * 获取节点的所有出边
     */
    public List<FlowEdge> getOutgoingEdges(TestFlow flow, String nodeId) {
        if (flow.getEdges() == null) return Collections.emptyList();
        return flow.getEdges().stream()
                .filter(e -> e.getSource().equals(nodeId))
                .collect(Collectors.toList());
    }

    /**
     * 获取节点的所有入边
     */
    public List<FlowEdge> getIncomingEdges(TestFlow flow, String nodeId) {
        if (flow.getEdges() == null) return Collections.emptyList();
        return flow.getEdges().stream()
                .filter(e -> e.getTarget().equals(nodeId))
                .collect(Collectors.toList());
    }

    /**
     * 根据ID获取节点
     */
    public FlowNode getNodeById(TestFlow flow, String nodeId) {
        if (flow.getNodes() == null) return null;
        return flow.getNodes().stream()
                .filter(n -> n.getId().equals(nodeId))
                .findFirst()
                .orElse(null);
    }

    /**
     * 查找 Start 节点
     */
    public FlowNode findStartNode(TestFlow flow) {
        if (flow.getNodes() == null) return null;
        return flow.getNodes().stream()
                .filter(n -> n.getType() == NodeType.Start)
                .findFirst()
                .orElse(null);
    }

    /**
     * 查找所有 End 节点
     */
    public List<FlowNode> findEndNodes(TestFlow flow) {
        if (flow.getNodes() == null) return Collections.emptyList();
        return flow.getNodes().stream()
                .filter(n -> n.getType() == NodeType.End)
                .collect(Collectors.toList());
    }
}
